File: DailyChallenges/CPP/leetCode_907.cpp
// 907. Sum of Subarray Minimums @ 2022/11/25

class Solution {
public:
    int sumSubarrayMins(vector<int>& arr) {
        int n = arr.size();
        vector<int> l(n, -1), stack;
        for (int i=0; i<n; ++i){
            while (!stack.empty() && arr[stack.back()]>arr[i])
                stack.pop_back();
            l[i] = (stack.empty()) ? (i+1) : (i-stack.back());
            stack.push_back(i);
        }
        
        long ans = 0L;
        stack.clear(); 
        for (int i=n-1; i>=0; --i){
            while (!stack.empty() && arr[stack.back()]>=arr[i])
                stack.pop_back();
            long r = (stack.empty()) ? (n-i) : (stack.back()-i);
            stack.push_back(i);
            ans = (ans + l[i]*r*arr[i]) % 1000000007;
        }
            
        return ans;
    }
};

File: DailyChallenges/Scala/LeetCode_907.scala
// 907. Sum of Subarray Minimums @ 2022/11/25

object Solution {
    def sumSubarrayMins(arr: Array[Int]): Int = {
        (0 until arr.length).foldLeft((List[Int]()), List[Int]())((p, i) => { 
            val s = p._2.dropWhile(arr(_)>arr(i)) 
            (i-s.headOption.getOrElse(-1)::p._1, i::s) // (ll, stack)
        })._1.foldLeft((0L, List[Int](), arr.length-1))((p, l) =>{
            val s = p._2.dropWhile(arr(_)>=arr(p._3))
            val r:Long = s.headOption.getOrElse(arr.length) - p._3
            ((p._1 + l*r*arr(p._3))% 1000000007, p._3::s, p._3-1) // (res, stack, idx)
        })._1.toInt
    }
}
